# Capstone project: E-commerce Analysis and Time Series Demand Forecasting
Janet Ting, DSIF-6

Notebook 1 of 4

<a id="toc"></a>
# Table of contents
1. **[Exceutive Summary](#executive)**
2. **[Background](#background)**
3. **[Problem Statement](#statement)**
4. **[Objective/ Performance Metrics](#objective)**
5. **[Datasets](#dataset)**
6. **[Data Merging & Cleaning](#clean)**
7. **[Data Dictionary](#dictionary)**
8. **[Outside Research](#research)**

<a id="executive"></a>
# [&#8629;](#toc) Executive Summary 


The key idea of this project is to use data science to help E-Commerce business owners (so-called `seller` in this dataset) to forecast the demand for their products or services. There are three goals for this project. Firstly, it is to better understand the data within the e-commerce site especially the sales order demand from the perspective of time series. Secondly, it is to design and build a time series forecasting model for sellers to forecast the sales order demand. Lastly, it is to create a forecast application to allow sellers to predict and visualize the sales demand. 

The data used was sourced from [Kaggle](https://www.kaggle.com/datasets/olistbr/brazilian-ecommerce),generously provided by Olist, the largest department store in Brazilian marketplaces. With purchase date information provided, this enables us to predict future sales demand using time-series. The performance metrics used to evaluate the models are RMSE and SMAPE.  

The exploratory analysis shows that E-Commece industry has upward trajectory trend in Brazil. Most orders were purchased on Monday and least on Saturday. Customers tend to purchase during 4-10pm of the day. Top most selling product catergories and their time-trend similar product categories were identified. The data is pre-processed on the purchases count and ran through two stationary tests - ADF and KPSS to check for stationarity. The best forecasting model, Exponential Smoothing forecasting model on Seasonal-Treand-based-on-Loess(STL) transformed data, showed RMSE of 67.18 and SMAPE of 21.36%. The forecasting webapp was built on Streamlit.

<a id="background"></a>
# [&#8629;](#toc) Background

[*E-commerce*](https://www.toppr.com/guides/business-environment/emerging-trends-in-business/electronic-commerce/) is a popular term for electronic commerce or even internet commerce. Over here, the seller exchanges data in the form of pistures, text, address for delievry etc. and then buyer make the payment. As of now, e-commerce is one of the fastest growing industries in the global economy. As per one estimate, it grows nearly 23% every year. And it's projected to be a [*$27 trillion*](https://www.insiderintelligence.com/insights/ecommerce-industry-statistics/) industry by the end of this decade or by 2020. 

In today's world, E-commerce industry is facing [*potential challenges*]() & [*limitations*](https://www.notifyvisitors.com/blog/limitations-of-ecommerce-business/). Just to name a few:
- Huge technological cost 
- Security
- Employee cost
- Investment/ Advertising cost
- Shipping cost 
- Packaging cost
- Warehouse storage cost 
- Marketing cost
- Complicated eCommerce policies
- Sales flow
- Customer Relationship Management(CRM) maintenance 

Therefore, [*demand and sales forecasting*](https://mobidev.biz/blog/machine-learning-methods-demand-forecasting-retail) are of paramount importance in retail and e-commerce. Without this tool, companies encounter disruption of the inventory balance, through ordering too much or not enough products for a certain period of time. In the case of surplus, a company is forced to offer discounts to sell products. Otherwise, it may face inventory issues. A shortage, in turn, results in lost profits. However, these problems can be solved by applying demand and sales forecasting to increase the return on inventory and determine the intention of future consumers to buy a specific product at a specific price.

<a id="statement"></a>
# [&#8629;](#toc) Problem Statement

At the era of promoting sustainability in business, organizations face this challenge along with ways to maximizing profit and reducing cost (e.g. investment cost, packaging cost, shipping cost, warehouse storage cost). The least of the situation a business wants is to overly produce a product or to turn down potential customers due to shortages. 

With the aim to enhance the services/ experience for sellers and grow sellers' group, I wish to leverage on this project:
- To better understand the databse within the e-commerce site, 
    - overall order demand over time
    - e-commerce's impact on economy
    - customer base by location (does this matter for e-commerce?)
    - which product is popular & in high demand
    - top selling product categories 
- To build a time series forecasting model for sellers to forecast the sales order demand up to 2 months, with the least RMSE and target SMAPE of 25%.

*Key Questions*: (supplement by [EDA](02_EDA_Feature_Engineering.ipynb))
1. What drive(s) the demand? 
    - Good `leadtime`/ delivery performance? 
    - Attractive and reasonable pricing? 
    - Quality products? 
    - Usability and practicability products? 
    - State-of-art products? 
    - Reasonating marketing strategies? 
    - During production promotion? 
1. Which product categories are trending and contributed the most to total products purchased?
1. Which cities do most products being purchased?
1. Is there any seasonality to the number of orders placed?
1. How do products reviews tell us about the order? 



<a id="objective"></a>
# [&#8629;](#toc) Objective/ Performance Metrics

Time-series forecasting model performance can be tested by evaluating a comparison between the actual values and the predicted values. In this project, two performance metrics, namely the root mean square error (RMSE), symmetric mean absolute percentage error (SMAPE) are used to evaluate the performance of each model.

When dealing with regression or forecasting problems, general metrics that we use are RMSE, MAE etc. RMSE, MAE are good if we are dealing with in small ranges and if the variation is very less. RMSE is helpful to get an idea of the magnitude of errors in terms of number of sales, but relative ones like SMAPE might be more interpretable.

In general use cases, we see data with lot of variation and when we look at RMSE metric, the values may not seem sensible. In such cases, we can go with MAPE or sMAPE. This is probably the single most commonly used forecasting metric in demand planning. This error measures expresses error as a percentage and can be used in evaluating models for different datasets.[Hamzaçebi, 2008](https://www.sciencedirect.com/science/article/abs/pii/S0020025508002958). In these two metrics, we are looking at how much forecast is deviating from actual in percentages. One of the drawbacks of using MAPE is that the values will be asymmetric, meaning, it is biased towards large values. 

<a id="dataset"></a>
# [&#8629;](#toc) Datasets

![](../assets/Datasets_schema.png)

Above is the schema of the datasets that I will using in this project. The dataset comes from Olist with almost two years of data and it’s only a sample of Olist’s actual orders.   
- `olist_orders_dataset.csv` : This table is connected to 4 other tables. It is used to connect all the details related to an order..
- `olist_order_items_dataset.csv` : It contains the details of an item that had been purchased such as shipping date, price and so on..
- `olist_order_reviews_dataset.csv`: It contains details related to any reviews posted by the customer on a particular product that he had purchased..
- `olist_products_dataset.csv` : It contains related to a product such as the ID, category name and measurements.
- `olist_order_payments_dataset.csv` : The information contained in this table is related to the payment details associated with a particular order.
- `olist_customers_dataset.csv` : Details the customer base information of this firm.
- `olist_geolocation_dataset.csv` : It contains geographical information related to both the sellers and customers.
- `olist_sellers_dataset.csv` : This table contains the information related to all the sellers who have registered with this firm.

## Import libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
# import matplotlib.pyplot as plt
# import seaborn as sns
# import os
# %matplotlib inline
# from matplotlib.gridspec import GridSpec

# import plotly.offline as py
# import plotly.express as px
# import plotly.graph_objs as go
# import json
# import requests
# import folium
# # import folium.folium.plugins as plugins
# # from folium.plugins import fast_marker_cluster, fullscreen, minimap, heat_map, heat_map_withtime
# from folium.plugins import FastMarkerCluster, Fullscreen, MiniMap, HeatMap, HeatMapWithTime, LocateControl

# Utilities
# from viz_utils import *

pd.set_option('display.max_columns', None)
pd.set_option('display.max_colwidth', None)

## Load datasets

In [2]:
# Load all datasets
customers_df = pd.read_csv('../datasets/olist_customers_dataset.csv')
geolocation_df = pd.read_csv('../datasets/olist_geolocation_dataset.csv')
order_items_df = pd.read_csv('../datasets/olist_order_items_dataset.csv')
order_payments_df = pd.read_csv('../datasets/olist_order_payments_dataset.csv')
order_reviews_df = pd.read_csv('../datasets/olist_order_reviews_dataset.csv')
orders_df = pd.read_csv('../datasets/olist_orders_dataset.csv')
products_df = pd.read_csv('../datasets/olist_products_dataset.csv')
sellers_df = pd.read_csv('../datasets/olist_sellers_dataset.csv')
category_name_df = pd.read_csv('../datasets/product_category_name_translation.csv')

In [3]:
# create list of tuples (df, df_name) for datasets
datasets = [(customers_df, 'customers'),
            (geolocation_df, 'geolocation'),
            (order_items_df, 'order_items'),
            (order_payments_df, 'order_payments'),
            (order_reviews_df, 'order_reviews'),
            (orders_df, 'orders'), 
            (products_df, 'products'),
            (sellers_df, 'sellers'),
            (category_name_df, 'category_name')]

In [4]:
def df_summary(df, df_name):
    """
    Summaries for all the variables in the dataset
    """
    print(df_name, 'dataset')
    print()
    
    # print shape 
    print(df.shape)
    print()
    
    # check for non-null and dtype information
    print("Non-null and Data Types")
    df.info()
    print()
    
    # check for duplicated rows
    print(f'Number of duplicated rows: {df[df.duplicated(keep=False)].shape[0]}')
    # check for any missing values, null or 'M' or '-'
    print(f"Number of missing values: {df.isin([np.nan, 'M', '-']).sum().sum()}")
    # check for any zero values
    print(f"Number of zero values: {df.isin(['0.0','0']).sum().sum()}")
    print()    
    
    # print columns with missing values, nan/ 'M'/ '-'
    print("Columns with missing values:")
    print(df.isin([np.nan,'M','-']).sum()[df.isin([np.nan,'M','-']).sum() > 0].sort_values(ascending=False))
    print()
    
    # print coumns with zero values
    print("Columns with zero values:")
    print(df.isin(['0.0', '0']).sum()[df.isin(['0.0', '0']).sum() > 0])
    print()
    
    print("______________________________________________________\n")

In [5]:
def first_five(df, df_name):
    """
    Show first 5 rows of the dataset
    """
    temp_df = df.head()
    print(f'Frist 5 rows of the Dataframe : {df_name} dataset')
    print()
    display(temp_df)

## Initial Analysis

### Customers dataset

In [6]:
df_summary(customers_df, 'customers')

customers dataset

(99441, 5)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB

Number of duplicated rows: 0
Number of missing values: 0
Number of zero values: 0

Columns with missing values:
Series([], dtype: int64)

Columns with zero values:
Series([], dtype: int64)

______________________________________________________



In [7]:
first_five(customers_df, 'customers')

Frist 5 rows of the Dataframe : customers dataset



customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [8]:
# Creating a DataFrame to get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset']= 5*[datasets[0][1]]
num_unique['features'] = datasets[0][0].columns
num_unique['n_unique'] = [datasets[0][0][col].nunique() for col in datasets[0][0].columns]
    
num_unique.style.background_gradient()

In [9]:
customers_df['customer_state'].value_counts()

SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: customer_state, dtype: int64

**Key notes:**
- We observed that the number of *customer_unique_id* is less than *customer_id*, suggesting customers use different id to link up their unique_id
- Customers base are from 27 states which consist of 4119 cities nationwide. 
- We could aggregate customers by state.

### Geolocation dataset

In [10]:
df_summary(geolocation_df, 'geolocation')

geolocation dataset

(1000163, 5)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000163 entries, 0 to 1000162
Data columns (total 5 columns):
 #   Column                       Non-Null Count    Dtype  
---  ------                       --------------    -----  
 0   geolocation_zip_code_prefix  1000163 non-null  int64  
 1   geolocation_lat              1000163 non-null  float64
 2   geolocation_lng              1000163 non-null  float64
 3   geolocation_city             1000163 non-null  object 
 4   geolocation_state            1000163 non-null  object 
dtypes: float64(2), int64(1), object(2)
memory usage: 38.2+ MB

Number of duplicated rows: 390005
Number of missing values: 0
Number of zero values: 0

Columns with missing values:
Series([], dtype: int64)

Columns with zero values:
Series([], dtype: int64)

______________________________________________________



In [11]:
first_five(geolocation_df, 'geolocation')

Frist 5 rows of the Dataframe : geolocation dataset



geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [12]:
# Creating a DataFrameto get unique count for each column
num_unique['dataset'] = 5*[datasets[1][1]]
num_unique['features'] = datasets[1][0].columns
num_unique['n_unique'] = [datasets[1][0][col].nunique() for col in datasets[1][0].columns]
    
num_unique.style.background_gradient()

**Key notes:**
- Brazil geolocation data has 27 states that consist of total 8011 cities.

### Order_Items dataset

In [13]:
df_summary(order_items_df, 'order_items')

order_items dataset

(112650, 7)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB

Number of duplicated rows: 0
Number of missing values: 0
Number of zero values: 0

Columns with missing values:
Series([], dtype: int64)

Columns with zero values:
Series([], dtype: int64)

______________________________________________________



In [14]:
first_five(order_items_df, 'order_items')

Frist 5 rows of the Dataframe : order_items dataset



order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [15]:
# Creating a DataFrameto get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset'] = 7*[datasets[2][1]]
num_unique['features'] = datasets[2][0].columns
num_unique['n_unique'] = [datasets[2][0][col].nunique() for col in datasets[2][0].columns]
    
num_unique.style.background_gradient()

**Key notes:**
- We observed that there are 32951 unique `product_id` with 5968 unique `price`

### Order_Payments dataset

In [16]:
df_summary(order_payments_df, 'order_payments')

order_payments dataset

(103886, 5)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  object 
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  object 
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB

Number of duplicated rows: 0
Number of missing values: 0
Number of zero values: 0

Columns with missing values:
Series([], dtype: int64)

Columns with zero values:
Series([], dtype: int64)

______________________________________________________



In [17]:
first_five(order_payments_df, 'order_payments')

Frist 5 rows of the Dataframe : order_payments dataset



order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [18]:
# Creating a DataFrameto get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset'] = 5*[datasets[3][1]]
num_unique['features'] = datasets[3][0].columns
num_unique['n_unique'] = [datasets[3][0][col].nunique() for col in datasets[3][0].columns]
    
num_unique.style.background_gradient()

**Key notes:**
- There are 5 unique `payment_type`s with 29 `payment sequential`
- 99440 unique `order_id` however only 29077 `payment_value` 

### Order_Reviews dataset

In [19]:
df_summary(order_reviews_df, 'order_reviews')

order_reviews dataset

(99224, 7)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   review_id                99224 non-null  object
 1   order_id                 99224 non-null  object
 2   review_score             99224 non-null  int64 
 3   review_comment_title     11568 non-null  object
 4   review_comment_message   40977 non-null  object
 5   review_creation_date     99224 non-null  object
 6   review_answer_timestamp  99224 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB

Number of duplicated rows: 0
Number of missing values: 145907
Number of zero values: 14

Columns with missing values:
review_comment_title      87657
review_comment_message    58250
dtype: int64

Columns with zero values:
review_comment_title      13
review_comment_message     1
dtype: int64

___________

In [20]:
first_five(order_reviews_df, 'order_reviews')

Frist 5 rows of the Dataframe : order_reviews dataset



review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                                                                                 review_comment_message  \
0                                                                                                   NaN   
1                                                                                                   NaN   
2                                                                                                   NaN   
3                                                                 Recebi bem antes do prazo estipulado.   
4  Parabéns lojas lannister adorei comprar pela Internet seguro e prático Parabéns a todos feliz Páscoa   

  review_creation_date review_answer_timestamp  
0  2018-01-18 00:00:00     2018-01-18 21:46:59  
1  2018-03-10 00:00:00     2018-03-11 03:05:13  
2  2018-02-17 00:00:00     2018-02-18 14:36:24  
3  2017-04-21 00:00:00     2017-04-21 22:02:06  
4  2018-03-01 00:00:00     2018-03-02 10:26:53

In [21]:
# Creating a DataFrameto get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset'] = 7*[datasets[4][1]]
num_unique['features'] = datasets[4][0].columns
num_unique['n_unique'] = [datasets[4][0][col].nunique() for col in datasets[4][0].columns]
    
num_unique.style.background_gradient()

**Key notes:**
- There are only 36159 unique review comment messages out of 98410 `review_id` and 98673 `order_id`
- Review creation date and review answer timestamp are of wrong data types, it should be datetime

### Orders dataset

In [22]:
df_summary(orders_df, 'orders')

orders dataset

(99441, 8)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB

Number of duplicated rows: 0
Number of missing values: 4908
Number of zero values: 0

Columns with missing values:
order_delivered_customer_date    2965
order_delivered_carrier_date     1783
order_approved_at 

In [23]:
first_five(orders_df, 'orders')

Frist 5 rows of the Dataframe : orders dataset



order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [24]:
# Creating a DataFrameto get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset'] = 8*[datasets[5][1]]
num_unique['features'] = datasets[5][0].columns
num_unique['n_unique'] = [datasets[5][0][col].nunique() for col in datasets[5][0].columns]
    
num_unique.style.background_gradient()

In [25]:
# return rows whereby the order_approved_at is null
orders_df[orders_df['order_approved_at'].isna() == 1]

order_id                       customer_id  \
1130   00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801   ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
1868   df8282afe61008dc26c6c31011474d02  aa797b187b5466bc6925aaaa4bb3bed1   
2029   8d4c637f1accf7a88a4555f02741e606  b1dd715db389a2077f43174e7a675d07   
2161   7a9d4c7f9b068337875b95465330f2fc  7f71ae48074c0cfec9195f88fcbfac55   
...                                 ...                               ...   
97696  5a00b4d35edffc56b825c3646a99ba9d  6a3bdf004ca96338fb5fad1b8d93c2e6   
98415  227c804e2a44760671a6a5697ea549e4  62e7477e75e542243ee62a0ba73f410f   
98909  e49e7ce1471b4693482d40c2bd3ad196  e4e7ab3f449aeb401f0216f86c2104db   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   

      order_status order_purchase_timestamp order_approved_at  \
1130      canceled      2018-08-28 15:26:39               NaN   
1801      canceled      2018-09-20 13:54:16               NaN   
1868      canceled      2017-03-04 12:14:30               NaN   
2029      canceled      2018-08-29 16:27:49               NaN   
2161      canceled      2017-05-01 16:12:39               NaN   
...            ...                      ...               ...   
97696     canceled      2017-07-02 15:38:46               NaN   
98415     canceled      2017-09-28 15:02:56               NaN   
98909     canceled      2018-08-07 11:16:28               NaN   
99283     canceled      2018-08-31 16:13:44               NaN   
99347     canceled      2018-09-06 18:45:47               NaN   

      order_delivered_carrier_date order_delivered_customer_date  \
1130                           NaN                           NaN   
1801                           NaN                           NaN   
1868                           NaN                           NaN   
2029                           NaN                           NaN   
2161                           NaN                           NaN   
...                            ...                           ...   
97696                          NaN                           NaN   
98415                          NaN                           NaN   
98909                          NaN                           NaN   
99283                          NaN                           NaN   
99347                          NaN                           NaN   

      order_estimated_delivery_date  
1130            2018-09-12 00:00:00  
1801            2018-10-17 00:00:00  
1868            2017-04-10 00:00:00  
2029            2018-09-13 00:00:00  
2161            2017-05-30 00:00:00  
...                             ...  
97696           2017-07-25 00:00:00  
98415           2017-10-16 00:00:00  
98909           2018-08-10 00:00:00  
99283           2018-10-01 00:00:00  
99347           2018-09-27 00:00:00  

[160 rows x 8 columns]

**Key notes:**
- Total 99441 `oders_id` while only 95664 orders delivered to customers `order_delivered_customer_date` successfully.
- Wrong data types for features:`order_purchase_timestamp`, `order_approved_at`, `order_delivered_carrier_date`, `order_delivered_customer_date`, `order_estimated_delivery_date`.

We can see that unlike `order_items`, orders_df contain some missing values for the columns `order_approved_at`, `order_delivered_carrier_date`, and `order_delivered_customer_date`. This is related to the topics 'Order Management' or 'Delivery and Fulfillment'. Having null values for `order_approved_at` makes sense since the payment for the order must go through successfully for it to be approved: two possible issues can be software or other tech issues with the payment processing, as well as card declines due to fraud prevention or lack of funds.

The fact that Olist does not offer logistics service aside from software at the moment where this data was collected. We can assume that they engage third party logistics companies for this. It's a bit unusual to see null values for `order_delivered_carrier_date` which I assume it tracks the moment the items of an order were handed to the logistics partners. We also see that `order_delivered_customer_date` has null values. Both cases might be related to orders being lost. Let's briefly check what's the percentage of these assumed lost orders from the total orders.

In [26]:
# calculate percentage of possible orders lost
assumed_lost = orders_df[orders_df['order_delivered_customer_date'].isnull() | orders_df['order_delivered_carrier_date'].isnull()]
assumed_lost['order_id'].count()/ orders_df['order_id'].count() * 100

2.9826731428686357

If the assumption that these orders were lost are correct, it accounts for 2.98% of the total orders which for the sake of our analysis will not cause issues.

### Products dataset

In [27]:
df_summary(products_df, 'products')

products dataset

(32951, 9)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  object 
 1   product_category_name       32341 non-null  object 
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), object(2)
memory usage: 2.3+ MB

Number of duplicated rows: 0
Number of missing values: 2448
Number of zero values: 0

Columns with missing values:
product_category_name         610
product_n

In [28]:
first_five(products_df, 'products')

Frist 5 rows of the Dataframe : products dataset



product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [29]:
# Creating a DataFrameto get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset'] = 9*[datasets[6][1]]
num_unique['features'] = datasets[6][0].columns
num_unique['n_unique'] = [datasets[6][0][col].nunique() for col in datasets[6][0].columns]
    
num_unique.style.background_gradient()

**Key notes:**
- There are many missing values for features: `product_category_name`,`product_name_lenght`,`product_description_lenght` and `product_photos_qty`. There were missing in random.

### Sellers dataset

In [30]:
df_summary(sellers_df, 'sellers')

sellers dataset

(3095, 4)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3095 entries, 0 to 3094
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   seller_id               3095 non-null   object
 1   seller_zip_code_prefix  3095 non-null   int64 
 2   seller_city             3095 non-null   object
 3   seller_state            3095 non-null   object
dtypes: int64(1), object(3)
memory usage: 96.8+ KB

Number of duplicated rows: 0
Number of missing values: 0
Number of zero values: 0

Columns with missing values:
Series([], dtype: int64)

Columns with zero values:
Series([], dtype: int64)

______________________________________________________



In [31]:
first_five(sellers_df, 'sellers')

Frist 5 rows of the Dataframe : sellers dataset



seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [32]:
# Creating a DataFrameto get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset'] = 4*[datasets[7][1]]
num_unique['features'] = datasets[7][0].columns
num_unique['n_unique'] = [datasets[7][0][col].nunique() for col in datasets[7][0].columns]
    
num_unique.style.background_gradient()

**Key notes:**
- Sellers are made up of 23 `seller_state`(s) and 611 `seller_city`(s), much less than customers group.

### Category dataset

In [33]:
df_summary(category_name_df, 'category_name')

category_name dataset

(71, 2)

Non-null and Data Types
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 71 entries, 0 to 70
Data columns (total 2 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   product_category_name          71 non-null     object
 1   product_category_name_english  71 non-null     object
dtypes: object(2)
memory usage: 1.2+ KB

Number of duplicated rows: 0
Number of missing values: 0
Number of zero values: 0

Columns with missing values:
Series([], dtype: int64)

Columns with zero values:
Series([], dtype: int64)

______________________________________________________



In [34]:
first_five(category_name_df, 'category_name')

Frist 5 rows of the Dataframe : category_name dataset



product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [35]:
# Creating a DataFrameto get unique count for each column
num_unique = pd.DataFrame({})
num_unique['dataset'] = 2*[datasets[8][1]]
num_unique['features'] = datasets[8][0].columns
num_unique['n_unique'] = [datasets[8][0][col].nunique() for col in datasets[8][0].columns]
    
num_unique.style.background_gradient()

**Key notes:**
- There are 71 unique `product_category_name_english`. 

<h3> Datasets Overview </h3>

In [36]:
# Creating a DataFrame with useful information about all datasets
# source code:[ThiagoPanini](https://github.com/ThiagoPanini/kaggle_challenges/tree/master/kernels/04_brazilian_e-commerce/notebooks)
data_info = pd.DataFrame({})
data_info['dataset'] = [datasets[n][1] for n in range(len(datasets))]
data_info['n_rows'] = [datasets[n][0].shape[0] for n in range(len(datasets))]
data_info['n_cols'] = [datasets[n][0].shape[1] for n in range(len(datasets))]
data_info['null_amount'] = [datasets[n][0].isnull().sum().sum() for n in range(len(datasets))] 
data_info['qty_null_columns'] = [len([col for col, null in datasets[n][0].isnull().sum().items() if null > 0]) for n in range(len(datasets))]
data_info['null_columns'] = [', '.join([col for col, null in datasets[n][0].isnull().sum().items() if null > 0]) for n in range(len(datasets))]

data_info.style.background_gradient()

The below homemade function enables us to look into each dataset and bring some detailed parameters about the data content. With this function we can get the following information for each dataset column:

- Column name;
- Null amount;
- Null percentage among the respective dataset;
- Data type;
- Total categorical entries;

In [37]:
def data_overview(df, corr=False, label_name=None, sort_by='qty_null', thresh_percent_null=0, thresh_corr_label=0):
    """
    Steps:
        1. Survey of attributes with null data in the set
        2. Analyze the primitive type of each attribute
        3. Analyze the number of the entries in categorical attribute 
        4. Extract the pearson correlation with the target for each attribute 
        5. Apply rules defined in the arguments
        6. Return the created overview dataset 

    Arguments:
        df -- DataFrame to parse [pandas.DataFrame]
        label_name -- target variable name [string]
        sort_by -- overview of dataset sort column [string - default: 'qty_null']
        thresh_percent_null -- Null data filter [int - default: 0]
        threh_corr_label -- correlation filter with target [int - default: 0]

    Return
        df_overview -- consolidated dataset containing analysis of columns [pandas.DataFrame]
    """

    # Creating Dataframe with null information
    df_null = pd.DataFrame(df.isnull().sum()).reset_index()
    df_null.columns = ['feature', 'qty_null']
    df_null['percent_null'] = df_null['qty_null'] / len(df)

    # Retuen primitive type and quantity of entries for categories
    df_null['dtype'] = df_null['feature'].apply(lambda x: df[x].dtype)
    df_null['qty_cat'] = [len(df[col].value_counts()) if df[col].dtype == 'object' else 0 for col in df_null['feature'].values]

    if corr:
        # Extract correlation information with target 
        label_corr = pd.DataFrame(df.corr()[label_name])
        label_corr = label_corr.reset_index()
        label_corr.columns = ['feature', 'target_pearson_corr']

        # Gather information 
        df_null_overview = df_null.merge(label_corr, how='left', on='feature')
        df_null_overview.query('target_pearson_corr > @thresh_corr_label')
    else:
        df_null_overview = df_null

    # Filter null data according to thresholds
    df_null_overview.query('percent_null > @thresh_percent_null')

    # Sort DataFrame
    df_null_overview = df_null_overview.sort_values(by=sort_by, ascending=False)
    df_null_overview = df_null_overview.reset_index(drop=True)

    return df_null_overview

In [38]:
# create DataFrame of all datasets summary
df_overview = pd.DataFrame({})
name_col = []
for df,df_name in datasets:
    name_col += [df_name] * df.shape[1]
    df_overview = df_overview.append(data_overview(df))   # data_overview function 
    df_overview['dataset_name'] = name_col

df_overview = df_overview.loc[:, ['dataset_name', 'feature', 'qty_null', 'percent_null', 'dtype', 'qty_cat']]
df_overview

dataset_name                        feature  qty_null  percent_null  \
0       customers                    customer_id         0      0.000000   
1       customers             customer_unique_id         0      0.000000   
2       customers       customer_zip_code_prefix         0      0.000000   
3       customers                  customer_city         0      0.000000   
4       customers                 customer_state         0      0.000000   
0     geolocation    geolocation_zip_code_prefix         0      0.000000   
1     geolocation                geolocation_lat         0      0.000000   
2     geolocation                geolocation_lng         0      0.000000   
3     geolocation               geolocation_city         0      0.000000   
4     geolocation              geolocation_state         0      0.000000   
0     order_items                       order_id         0      0.000000   
1     order_items                  order_item_id         0      0.000000   
2     order_items                     product_id         0      0.000000   
3     order_items                      seller_id         0      0.000000   
4     order_items            shipping_limit_date         0      0.000000   
5     order_items                          price         0      0.000000   
6     order_items                  freight_value         0      0.000000   
0  order_payments                       order_id         0      0.000000   
1  order_payments             payment_sequential         0      0.000000   
2  order_payments                   payment_type         0      0.000000   
3  order_payments           payment_installments         0      0.000000   
4  order_payments                  payment_value         0      0.000000   
0   order_reviews           review_comment_title     87656      0.883415   
1   order_reviews         review_comment_message     58247      0.587025   
2   order_reviews                      review_id         0      0.000000   
3   order_reviews                       order_id         0      0.000000   
4   order_reviews                   review_score         0      0.000000   
5   order_reviews           review_creation_date         0      0.000000   
6   order_reviews        review_answer_timestamp         0      0.000000   
0          orders  order_delivered_customer_date      2965      0.029817   
1          orders   order_delivered_carrier_date      1783      0.017930   
2          orders              order_approved_at       160      0.001609   
3          orders                       order_id         0      0.000000   
4          orders                    customer_id         0      0.000000   
5          orders                   order_status         0      0.000000   
6          orders       order_purchase_timestamp         0      0.000000   
7          orders  order_estimated_delivery_date         0      0.000000   
0        products          product_category_name       610      0.018512   
1        products            product_name_lenght       610      0.018512   
2        products     product_description_lenght       610      0.018512   
3        products             product_photos_qty       610      0.018512   
4        products               product_weight_g         2      0.000061   
5        products              product_length_cm         2      0.000061   
6        products              product_height_cm         2      0.000061   
7        products               product_width_cm         2      0.000061   
8        products                     product_id         0      0.000000   
0         sellers                      seller_id         0      0.000000   
1         sellers         seller_zip_code_prefix         0      0.000000   
2         sellers                    seller_city         0      0.000000   
3         sellers                   seller_state         0      0.000000   
0   category_name          product_category_name         0      0.000000   
1   category_name  product_category_name_english     

Section summary:
1. The column 'review_comment_title' has the most missing value, it's probably due to the fact that customers usually leave comment messages without title in an e-Commerce platform.
2. Customers are not compulsory to leave a review upon receiving the goods and they have no incentive to do so. So, 58.7% of 'review_comment_message' is null.
3. The missing values in columns 'order_delivered_customer_date', 'order_delivered_carrier_date' and 'order_approved_at' reflect that those orders that are not completedly delivered (i.e. order_status is not 'delivered')
4. Products with missing attributes, such as product_category_name, product_name_lenght, description_lenght and photos_qty, are purely missing in random.

<a id="clean"></a>
# [&#8629;](#toc) Data Merging & Cleaning

In [39]:
# Merging datasets mainly by 'left'(takes in primary df), 'outer'(includes everything from all df)
df_orders = orders_df.merge(customers_df, how='left', on='customer_id')
df_orders_items = df_orders.merge(order_items_df, how='left', on='order_id')
df_train = df_orders_items.merge(order_payments_df, how='outer', on='order_id', validate='m:m').merge(order_reviews_df, how='outer', on='order_id')
df_train = df_train.merge(products_df, how='outer', on='product_id').merge(category_name_df, how='outer', on='product_category_name')
df_train = df_train.merge(sellers_df, how='outer', on='seller_id')
print(df_orders_items.shape)
print(df_train.shape)

(113425, 18)
(119143, 40)


In [40]:
# correct the datatype for date columns
timestamp_cols = ['order_purchase_timestamp', 'order_approved_at', 'order_estimated_delivery_date', 'order_delivered_customer_date', 'shipping_limit_date' ,'review_creation_date', 'review_answer_timestamp']
for col in timestamp_cols:
    df_train[col] = pd.to_datetime(df_train[col])

In [41]:
# save to csv
df_orders_items.to_csv('../datasets/orders_items.csv')
df_train.to_csv('../datasets/olist.csv')

In [42]:
df_orders_items.head(3)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date                customer_unique_id  \
0           2017-10-18 00:00:00  7c396fd4830fd04220f754e42b4e5bff   
1           2018-08-13 00:00:00  af07308b275d755c9edb36a90c618231   
2           2018-09-04 00:00:00  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix customer_city customer_state  order_item_id  \
0                      3149     sao paulo             SP            1.0   
1                     47813     barreiras             BA            1.0   
2                     75265    vianopolis             GO            1.0   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  595fac2a385ac33a80bd5114aec74eb8  289cdb325fb7e7f891c38608bf9e0962   
2  aa4383b373c6aca5d8797843e5594415  4869f7a5dfa277a7dca6462dcf3b52b2   

   shipping_limit_date   price  freight_value  
0  2017-10-06 11:07:15   29.99           8.72  
1  2018-07-30 03:24:27  118.70          22.76  
2  2018-08-13 08:55:23  159.90          19.22

In [43]:
df_train.head(3)

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   

  order_estimated_delivery_date                customer_unique_id  \
0                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   
1                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   
2                    2017-10-18  7c396fd4830fd04220f754e42b4e5bff   

   customer_zip_code_prefix customer_city customer_state  order_item_id  \
0                      3149     sao paulo             SP            1.0   
1                      3149     sao paulo             SP            1.0   
2                      3149     sao paulo             SP            1.0   

                         product_id                         seller_id  \
0  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
1  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   
2  87285b34884572647811a353c7ac498a  3504c0cb71d7fa48d967e0e4c94d59d9   

  shipping_limit_date  price  freight_value  payment_sequential payment_type  \
0 2017-10-06 11:07:15  29.99           8.72                 1.0  credit_card   
1 2017-10-06 11:07:15  29.99           8.72                 3.0      voucher   
2 2017-10-06 11:07:15  29.99           8.72                 2.0      voucher   

   payment_installments  payment_value                         review_id  \
0                   1.0          18.12  a54f0611adc9ed256b57ede6b6eb5114   
1                   1.0           2.00  a54f0611adc9ed256b57ede6b6eb5114   
2                   1.0          18.59  a54f0611adc9ed256b57ede6b6eb5114   

   review_score review_comment_title  \
0           4.0                  NaN   
1           4.0                  NaN   
2           4.0                  NaN   

                                                                                                                                                       review_comment_message  \
0  Não testei o produto ainda, mas ele veio correto e em boas condições. Apenas a caixa que veio bem amassada e danificada, o que ficará chato, pois se trata de um presente.   
1  Não testei o produto ainda, mas ele veio correto e em boas condições. Apenas a caixa que veio bem amassada e danificada, o que ficará chato, pois se trata de um presente.   
2  Não testei o produto ainda, mas ele veio correto e em boas condições. Apenas a caixa que veio bem amassada e danificada, o que ficará chato, pois se trata de um presente.   

  review_creation_date review_answer_timestamp  product_category_name  \
0           2017-10-11     2017-10-12 03:43:48  utilidades_domesticas   
1           2017-10-11     2017-10-12 03:43:48  utilidades_domesticas   
2           2017-10-11     2017-10-12 03:43:48  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       268.0                 4.0   
1                 40.0                       268.0                 4.0   
2                 40.0                       268.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  \
0             500.0               19.0                8.0              13.0   
1             500.0               19.0                8.0              13.0   
2             500.0           

In [44]:
# potential cleaning 
# discover that 73 orders are approved after customers have received the goods, possibly that was the effect of [fraud-protection software](https://www.bolt.com/thinkshop/order-approval-rate-for-ecommerce)
df_late_approval = df_train[df_train['order_approved_at'] > df_train['order_delivered_customer_date']]
print(df_late_approval.shape)
df_late_approval

(73, 40)


order_id                       customer_id  \
1890    6b80bb20190715d71c43efff617bd659  2fedecfd993b8b3fa889d00eee230748   
3698    0184d4ddb259e1a4cfc2871888cf97b8  09425ea1839abf2f0d289a0ff453fa21   
3773    66e1b657a71397245290f39ffe24031e  be274b149cc9bb8af13628e268867c10   
3774    66e1b657a71397245290f39ffe24031e  be274b149cc9bb8af13628e268867c10   
4590    f222c56f035b47dfa1e069a88235d730  b74ca180d63f9ae0443e4e13a2f5bdaf   
...                                  ...                               ...   
114945  3a41d0e227b4c6fff055561af5eaca13  345a1ac3ca53330b45af176810ac91a6   
115847  ecd1661658d34312cf9ac2bf34ba3b6b  bcec3938d91ba9a3bf2c4261a44cb9f3   
115910  547b3596c03dfad1b87305e9c5a554c2  e14b08a5e109d77846ce7d4cb8b12d6c   
115911  547b3596c03dfad1b87305e9c5a554c2  e14b08a5e109d77846ce7d4cb8b12d6c   
117514  4f3a6e28d764cf896b1fceb0028422c8  e24fceba7ac4b1b1cd1884f05ba68e9d   

       order_status order_purchase_timestamp   order_approved_at  \
1890      delivered      2017-02-19 01:15:03 2017-03-01 10:51:46   
3698      delivered      2017-09-01 20:04:28 2017-09-13 22:17:15   
3773      delivered      2018-07-22 15:09:37 2018-07-27 23:31:10   
3774      delivered      2018-07-22 15:09:37 2018-07-27 23:31:10   
4590      delivered      2018-01-30 09:43:45 2018-02-04 23:31:47   
...             ...                      ...                 ...   
114945    delivered      2018-07-03 15:00:40 2018-07-05 16:11:06   
115847    delivered      2018-06-30 18:49:36 2018-07-05 16:25:45   
115910    delivered      2018-04-22 23:05:38 2018-04-24 19:23:54   
115911    delivered      2018-04-22 23:05:38 2018-04-24 19:23:54   
117514    delivered      2018-07-03 09:34:16 2018-07-05 16:21:50   

       order_delivered_carrier_date order_delivered_customer_date  \
1890            2017-02-22 16:05:29           2017-02-24 14:27:26   
3698            2017-09-04 14:05:50           2017-09-09 15:12:44   
3773            2018-07-24 19:07:00           2018-07-25 14:58:47   
3774            2018-07-24 19:07:00           2018-07-25 14:58:47   
4590            2018-01-31 19:48:44           2018-02-01 20:10:38   
...                             ...                           ...   
114945          2018-07-04 15:11:00           2018-07-05 14:16:41   
115847          2018-07-03 11:46:00           2018-07-05 12:58:37   
115910          2018-04-23 19:04:05           2018-04-24 19:18:35   
115911          2018-04-23 19:04:05           2018-04-24 19:18:35   
117514          2018-07-03 16:52:00           2018-07-05 13:32:49   

       order_estimated_delivery_date                customer_unique_id  \
1890                      2017-03-17  4687b78e990956109c4019200e3ac062   
3698                      2017-09-20  10c573ba276f9857dc6820e0dde57fb8   
3773                      2018-08-01  3e3d56f9737c010ccb5b67c601a18b53   
3774                      2018-08-01  3e3d56f9737c010ccb5b67c601a18b53   
4590                      2018-02-15  398957fce766920db4c4b206d7c9aa85   
...                              ...                               ...   
114945                    2018-07-16  e7c3d1c547decbf4755ea2bac75b314d   
115847                    2018-07-30  6cc12c7fe99ee89f0c685359fa81868a   
115910                    2018-05-10  799f96a722913e43fa42d9b0b49ae861   
115911                    2018-05-10  799f96a722913e43fa42d9b0b49ae861   
117514                    2018-07-20  7410c7ed56c93d4eee17a7b7b3a75b3e   

        customer_zip_code_prefix customer_city customer_state  order_item_id  \
1890                       19906      ourinhos             SP            1.0   
3698                       14811    araraquara             SP            1.0   
3773                        7130     guarulhos             SP            1.0   
3774                        7130     guarulhos             SP            2.0   
4590                        9271   santo andre             SP            1.0   
...                          ...           ...            ...            ...   


In [45]:
# count number of orders with late approval
df_late_approval['order_status'].value_counts()

delivered    73
Name: order_status, dtype: int64

<a id="dictionary"></a>
# [&#8629;](#toc) Data Dictionary

| **Columns** | **Type** | **Dataset** | **Description** |
|---|---|---|---|
| **order_id** | *object* | olist_orders/ olist_order_items/ olist_order_payment | The id of the orders |
| **customer_id/customer_unique_id** | *object* | olist_order/ olist_customers | The id of the customers |
| **order_status** | *object* | olist_order | The status of the orders |
| **order_purchase_timestamp** | *datetime* | olist_order | The timestamp when the purchase orders are generated |
| **order_approved_at** | *datetime* | olist_order | The timestamp when the orders are approved |
| **order_delivered_carrier_date** | *datetime* | olist_order | The timestamp when the orders are delivered to carriers |
| **order_delivered_customer_date** | *datetime* | olist_order | The timestamp when the orders are delivered to customers |
| **order_estimated_delivery_date** | *datetime* | olist_order | The estimated date when the orders are delivered to customers |
| **customer_zip_code_prefix** | *integer* | olist_customers | The zip code prefix of the customers |
| **customer_city** | *object* | olist_customers | The city of the customers |
| **customer_state** | *object* | olist_customers | The state of the customers |
| **order_items_id** | *integer* | olist_order_items | The id of the items within orders |
| **product_id** | *object* | olist_order_items | The id of the products |
| **seller_id** | *object* | olist_order_items | The id of the sellers |
| **shipping_limit_date** | *datetime* | olist_order_items | The timestamp the company estimates the ordered products will be shipped |
| **price** | *float* | olist_order_items | The price of the orders |
| **freight_value** | *float* | olist_order_items | The freight of the orders |
| **payment_sequential** | *integer* | olist_order_payments | The sequence of the order payments |
| **payment_type** | *object* | olist_order_payments | The type of the order payments |
| **payment_installments** | *integer* | olist_order_payments | The number of installments the order payments |
| **payment_value** | *float* | olist_order_payments | The value of the order payments |
| **review_id** | *object* | olist_order_reviews | The id of the order reviews |
| **review_score** | *integer* | olist_order_reviews | The reviews score of the order given by customers |
| **review_comment_title** | *object* | olist_order_reviews | The title of the order reviews |
| **review_comment_message** | *object* | olist_order_reviews | The message content of the order reviews |
| **review_creation_date** | *datetime* | olist_order_reviews | The datetime when the reviews are created by customers |
| **review_answer_timestamp** | *datetime* | olist_order_reviews | The datetime when the reviews are answered by sellers  |
| **product_category_name** | *object* | olist_products | The name of product categories in Portuguese |
| **product_name_lenght** | *integer* | olist_products | The length of product names in Portuguese |
| **product_description_lenght** | *object* | olist_products | The length of product descriptions in Portuguese |
| **product_photos_qty** | *object* | olist_products | The number of product photos |
| **product_weight_g** | *object* | olist_products | The weight of the products in grams |
| **product_length_cm** | *integer* | olist_products | The length of products (in centimeters) |
| **product_height_cm** | *object* | olist_products | The height of products (in centimeters) |
| **product_width_cm** | *object* | olist_products | The width of products (in centimeters)|
| **product_category_name_english** | *object* | olist_products | The name of product categories translated to English |
| **seller_zip_code_prefix** | *integer* | olist_sellers | The zip code prefix of the sellers |
| **seller_city** | *object* | olist_sellers | The city of the sellers |
| **seller_state** | *object* | olist_sellers | The state of the sellers |

<a id="research"></a>
# [&#8629;](#toc) Outside Research

1. [*Market demand forecasting*](https://prevedere.com/3-real-world-market-demand-forecasting-stories/) is a critical process for any business, but perhaps none more so than those in consumer packaged goods. Inventory, production, storage, shipping, marketing – every facet of CPG and retail companies’ operations are affected by accurate forecasting. Identifying consumers’ preferences and their likelihood to purchase enables these companies to make better decisions regarding product lines, entering new markets and their supply chains, ensure that shelves are stocked, and minimize the risk of inventory shortages or overages. There are three real world market demand forecasting stories -[Beverage Brand, Electronics Company, Convenience Store](https://prevedere.com/3-real-world-market-demand-forecasting-stories/) 


2. [*Demand forecasts*](https://backinstock.org/blog/ecommerce-demand-forecasting/) should account for consumer purchase behavior, as well as your customers’ past purchase patterns.For each of your products and services across all sales channels over the past 12 months (or as close as possible), it’s best practice to identify their:
    - Order dates,
    - Rates of returns,
    - Monthly sales averages,
    - Moving sales averages, and
    - Underlying and seasonal market conditions or trends. 
    
    <u>Time-series analysis</u>
    Time-series analysis uses past sales data to draw conclusions about events and seasonality throughout the year. When we analyze exactly when a product or service sold well (such as a jacket around Black Friday), we can safely assume the pattern will repeat next year. Based on that, we can logically adjust your marketing tactics and stock levels.

    <u>Causal forecasting</u>
    Causal forecasting uses as much information as possible to predict accurate sales forecasts over the upcoming year. This includes data on sales, competitors, marketing activity, economic conditions, and even weather.

    It’s easy and convenient to get our ecommerce demand forecasting done with a solid tool like **Back In Stock**.

3. [The five basic steps of demand forecasting?](https://www.espeedpost.in/blog/what-are-the-five-basic-steps-of-demand-forecasting/)
    1. Identification of Objective.
    1. Nature of Product and Market.
    1. Determinants of Demand.
    1. Analysis of Factors.
    1. Choice of Method.
    1. Testing Accuracy.

4. [Amazon is reiventing Forecasting with Machine Learning](https://www.forbes.com/sites/amazonwebservices/2021/12/03/predicting-the-future-of-demand-how-amazon-is-reinventing-forecasting-with-machine-learning/?sh=53cb19661b6b) What if you could use data to predict what a customer will buy, one year before they even know they want it? How Amazon.com became a leader in product forecasting **According to Freshwater, Amazon’s journey with machine learning began about 10 years ago to improve forecast accuracy. “We started to use machine learning because our moving average models were just not as accurate as we had hoped they would be.”**


5. [Demand forecasting](https://tryolabs.com/blog/2020/06/01/price-optimization-for-e-commerce-a-case-study#4-modeling-the-demand) estimate the demand curves for each one of the items, i.e., understand the elasticity of the demand. That was affected by several factors (such as the price, competitor's price, season, holidays, and macroeconomic variables).


6. The [Olist](https://olist.com/) is a growing, successful, and customer-centric Brazilian based platform that connects different established e-commerce marketplaces and gives entrepreneurs and business owners the avenue to sell to a wide range of customers. It is a **Brazilian** departmental store (*marketplace*) that **operates in e-commerce segment, but is not an e-commerce itself** (*as she says*). Since 2015, it operates as a SaaS (*Software as a Service*) technology company that helps merchants, particularly brick-and-mortar ones, to sell their products in different online and physical marketplaces. Olist provides technology to their sellers to optimize the way their products are placed in different online marketplaces, and to track their performance across all the marketplaces where they are being sold. 

    Olist's solution consists of three aspects: Software, Contracts with the main marketplaces and Reputation sharing. The diagram below shows how Olist links marketplaces, consumers (*Portuguese: consumidores*), and retailers (*Portuguese: varejistas*)

    <b>What Olist says?</b>

    Olist says she... 

    1. ... is a large department store within marketplaces.
    2. ... is connected to the main e-commerces of Brazil.
    3. ... does not buy products. 
    4. ... does not keep products in stock.
    5. ... does not carry out shipping of any products offered in its store. 
    6. All products are sold and shipped by the thousands of shopkeepers (**registered on Olist***) who sell through Olist.
    7. Her strength lies in union of all participating shopkeepers, who are selling physical products. 
    8. Participant shopkeeper is responsible for separating, packing, and taking products to the logistics operator.

    **Please note Olist's perspective (a supply chain preview):** she prescribes there are many factors that can influence the sales of a shopkeeper e.g. type of product, demand, seasonality, competitive pricing, terms, inventory etc


7. [Goldman Sachs leads $23M in funding for Brazilian e-commerce startup Olist*](https://techcrunch.com/2021/04/15/goldman-sachs-leads-23m-in-funding-for-brazilian-e-commerce-startup-olist/)

8. [Types of demand forecasting planning softwares for e-commerce](https://www.inventory-planner.com/10-best-demand-forecasting-planning-software-for-e-commerce/) tools available on market. 

9. [$1.8 billion loss in productivity](https://financesonline.com/productivity-statistics/) 

**Key notes:** 
This notebook mainly produces preliminary analysis on all the datasets and summarizes each columns unique count. Through this, we gain an rough idea of what to expect in subsequent steps. 